feat!: Support compression codecs for Avro Files (including manifest and manifest lists)

emkornfield opened the pull request:
## Which issue does this PR close?


- Partially fixes #1731

## What changes are included in this PR?

Previously these properties where not honored on table properties.

- Adds table properties for these values.
- Plumbs them through for writers.

AI was used to help generate tests and some of the plumbing.

## Are these changes tested?

Added unit tests

BREAKING CHANGE: Adds codec parameter to some public functions.  By default start compressing manifests and manifest lists.

emkornfield committed:
feat!: Support compression codecs for JSON metadata and Avro

Previously these properties where not honored on tabel properties.

- Adds table properties for these values.
- Plumbs them through for writers.

kevinjqliu commented:
looks like a clippy issue in CI
```
error: this function has too many arguments (8/7)
   --> crates/iceberg/src/spec/manifest_list.rs:243:5
    |
243 | /     fn new(
244 | |         format_version: FormatVersion,
245 | |         output_file: OutputFile,
246 | |         metadata: HashMap<String, String>,
...   |
251 | |         compression_level: u8,
252 | |     ) -> Self {
    | |_____________^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#too_many_arguments
    = note: `-D clippy::too-many-arguments` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(clippy::too_many_arguments)]`
```

emkornfield committed:
fix clippy

emkornfield committed:
clippy again

emkornfield commented:
@liurenjie1024 @kevinjqliu would you have time to review?

kevinjqliu commented on crates/iceberg/src/spec/table_metadata.rs:
```diff
@@ -461,9 +461,57 @@ impl TableMetadata {
+                // Modify filename to add .gz before .metadata.json
+                let location = metadata_location.as_ref();
+                let new_location = if location.ends_with(".metadata.json") {
+                    location.replace(".metadata.json", ".gz.metadata.json")
```
😄  https://iceberg.apache.org/spec/#naming-for-gzip-compressed-metadata-json-files

i wonder if its a good idea to just default to `metadata.json.gz` here and not create more files with the `.gz.metadata.json` suffix

kevinjqliu reviewed (commented):
Generally LGTM, I left a few comments around using the same string as the java implementation

emkornfield commented on crates/iceberg/src/transaction/snapshot.rs:
```diff
@@ -205,7 +220,9 @@ impl<'a> SnapshotProducer<'a> {
                 .clone(),
-        );
+        )
+        .with_compression(compression);
```
I don't know that there is strong reason one way or another, happy to do whatever reviewers prefer.  As an argument against adding another parameter to the construct, Rust has clippy warnings for too many parameters, so for optional parameters it seems best not to add them?

emkornfield replied:
I'll see if I can make it all with constructors to make it consistent.

emkornfield replied:
This also fixes the bug mentioned below.

emkornfield commented on crates/iceberg/src/spec/manifest_list.rs:
```diff
@@ -198,23 +202,52 @@ impl ManifestListWriter {
+        // Copy over existing metadata from the old writer
+        // Unfortunately, we can't extract metadata from the old writer,
+        // so we'll need to handle this differently
+        self.avro_writer = new_writer;
```
Yeah, left over TODO, i'll fix this.

emkornfield committed:
wip

emkornfield committed:
address comments

emkornfield commented on crates/iceberg/src/spec/avro_util.rs:
```diff
@@ -0,0 +1,280 @@
+        }
+        Some("zstd") => {
+            // Zstandard supports levels 0-22, clamp to valid range
+            let zstd_level = level.min(22);
```
made this default to 1 for ZSTD

emkornfield commented on crates/iceberg/src/spec/avro_util.rs:
```diff
@@ -0,0 +1,280 @@
+///
+/// - `gzip` or `deflate`: Uses Deflate compression with specified level
+/// - `zstd`: Uses Zstandard compression (level clamped to valid zstd range)
+/// - `none` or `None`: No compression
```
fixed.

emkornfield commented on crates/iceberg/src/spec/avro_util.rs:
```diff
@@ -0,0 +1,308 @@
+///
+/// # Compression Levels
+///
+/// The compression level mapping is based on miniz_oxide's CompressionLevel enum:
```
clarify this is for gzip.

kevinjqliu replied:
nit: maybe merge this comment with the one above (L61-L66), the compression levels are explained twice

emkornfield committed:
no clone needed

emkornfield committed:
test compression works

kevinjqliu commented on crates/iceberg/src/spec/table_properties.rs:
```diff
@@ -137,6 +162,21 @@ impl TableProperties {
+    /// Compression codec for metadata files (JSON)
+    pub const PROPERTY_METADATA_COMPRESSION_CODEC: &str = "write.metadata.compression-codec";
+    /// Default metadata compression codec - uncompressed
+    pub const PROPERTY_METADATA_COMPRESSION_CODEC_DEFAULT: &str = "uncompressed";
```
```suggestion
    /// Default metadata compression codec - uncompressed
    pub const PROPERTY_METADATA_COMPRESSION_CODEC_DEFAULT: &str = "none";
```
this should be none according to 
https://iceberg.apache.org/docs/nightly/configuration/#write-properties

<img width="769" height="68" alt="Image" src="https://github.com/user-attachments/assets/75c8885b-4653-43ce-978b-6573646f6e69" />

kevinjqliu commented on crates/iceberg/src/spec/table_metadata.rs:
```diff
@@ -461,9 +461,57 @@ impl TableMetadata {
+            .unwrap_or(crate::spec::table_properties::TableProperties::PROPERTY_METADATA_COMPRESSION_CODEC_DEFAULT);
+
+        let (data_to_write, actual_location) = match codec {
+            "gzip" => {
```
nit: make this comparison case insensitive 

https://github.com/apache/iceberg/blob/700575f6e58c655ba680a0e5cc25d5a88c225d3c/core/src/main/java/org/apache/iceberg/TableMetadataParser.java#L63

kevinjqliu commented on crates/iceberg/src/transaction/snapshot.rs:
```diff
@@ -410,6 +444,7 @@ impl<'a> SnapshotProducer<'a> {
                 self.table.metadata().current_snapshot_id(),
                 next_seq_num,
                 Some(first_row_id),
+                compression,
```
```suggestion
                compression.clone(),
```
should this be .clone?  or should the rest be without .clone?

kevinjqliu commented on crates/iceberg/src/spec/manifest/writer.rs:
```diff
@@ -410,7 +423,11 @@ impl ManifestWriter {
         };
-        let mut avro_writer = AvroWriter::new(&avro_schema, Vec::new());
+
+        // Determine compression codec using CompressionSettings
```
```suggestion
```

kevinjqliu commented on crates/iceberg/src/spec/avro_util.rs:
```diff
@@ -0,0 +1,308 @@
+///
+/// # Arguments
+///
+/// * `codec` - The name of the compression codec (e.g., "gzip", "zstd", "deflate", "uncompressed")
```
is it easy to add snappy to completely match all the supported codex in java?

kevinjqliu reviewed (commented):
this looks great, thanks!
adding the "breaking" label since we're modifying a few public functions

emkornfield committed:
comments

emkornfield committed:
update tests

liurenjie1024 commented on crates/iceberg/src/spec/table_properties.rs:
```diff
@@ -34,6 +34,25 @@ where
+fn parse_optional_property<T: std::str::FromStr>(
+    properties: &HashMap<String, String>,
+    key: &str,
+) -> Result<Option<T>, anyhow::Error>
```
We should use this crate's error.

liurenjie1024 reviewed (changes requested):
Thanks @emkornfield for this pr. In general, I think this is a useful feature implementation. But we should not mix two things in one pr, I would suggest to split them into two prs: one for table metadata, and one for manifests.

emkornfield committed:
address comments

emkornfield committed:
remove parse optional property

emkornfield commented on crates/iceberg/src/spec/table_metadata.rs:
```diff
@@ -461,9 +461,59 @@ impl TableMetadata {
     ) -> Result<()> {
+        use std::io::Write as _;
+
+        use flate2::write::GzEncoder;
```
Still new to rust, I thought there was some convention to try to limit scope for rare imports.  I've moved this to the top.

emkornfield committed:
fmt

kevinjqliu commented on crates/iceberg/src/spec/table_properties.rs:
```diff
@@ -175,6 +196,19 @@ impl TryFrom<&HashMap<String, String>> for TableProperties {
+                    255u8,
+                )?;
+                if level == 255 { None } else { Some(level) }
+            },
```
is this not valid here? 
```
            avro_compression_level: parse_property(
                props,
                TableProperties::PROPERTY_AVRO_COMPRESSION_LEVEL,
                TableProperties::PROPERTY_AVRO_COMPRESSION_LEVEL_DEFAULT,
            )?,
```

`avro_compression_level` and `PROPERTY_AVRO_COMPRESSION_LEVEL_DEFAULT` are both `Option<u8>`

emkornfield replied:
I might be misunderstanding the comment.  But  I should have cleaned up `pub const PROPERTY_AVRO_COMPRESSION_LEVEL_DEFAULT: Option<u8> = None`;, I don't think it is used anymore.

In a prior revision I added a method (`parse_optional_property`) that could return  `None` value directly but @liurenjie1024 questioned whether this was useful.

kevinjqliu commented on crates/iceberg/src/spec/avro_util.rs:
```diff
@@ -0,0 +1,310 @@
+    fn default() -> Self {
+        Self {
+            codec: TableProperties::PROPERTY_AVRO_COMPRESSION_CODEC_DEFAULT.to_string(),
+            level: None,
```
```suggestion
            level: TableProperties::PROPERTY_AVRO_COMPRESSION_LEVEL_DEFAULT,
```

emkornfield replied:
removing this.

kevinjqliu commented on crates/iceberg/src/spec/table_properties.rs:
```diff
@@ -49,6 +56,10 @@ pub struct TableProperties {
     pub write_target_file_size_bytes: usize,
+    /// Compression codec for Avro files (manifests, manifest lists)
+    pub avro_compression_codec: String,
+    /// Compression level for Avro files (None uses codec-specific defaults: gzip=9, zstd=1)
```
```suggestion
    /// Compression level for Avro files (None uses codec-specific defaults)
```
nit: these might change in the future, so i think its better to not specify the values here 

there are a few other instances of this same comment

kevinjqliu reviewed (approved):
This LGTM, a couple of nit comments

Thanks for adding the additional tests for manifest and manifest list writers
I also added the "breaking" label, forgot to after i commented 😄

liurenjie1024 commented on crates/iceberg/src/spec/avro_util.rs:
```diff
@@ -0,0 +1,310 @@
+
+    // Use case-insensitive comparison to match Java implementation
+    match codec.map(|s| s.to_lowercase()).as_deref() {
+        Some("gzip") => {
```
Use constants rather than raw string.

liurenjie1024 commented on crates/iceberg/src/spec/table_properties.rs:
```diff
@@ -49,6 +56,10 @@ pub struct TableProperties {
     /// The target file size for files.
     pub write_target_file_size_bytes: usize,
+    /// Compression codec for Avro files (manifests, manifest lists)
+    pub avro_compression_codec: String,
```
```suggestion
    pub avro_compression_codec: Codec,
```

liurenjie1024 commented on crates/iceberg/src/spec/table_properties.rs:
```diff
@@ -49,6 +56,10 @@ pub struct TableProperties {
+    /// Compression codec for Avro files (manifests, manifest lists)
+    pub avro_compression_codec: String,
+    /// Compression level for Avro files (None uses codec-specific defaults: gzip=9, zstd=1)
+    pub avro_compression_level: Option<u8>,
```
We don't need this.

liurenjie1024 commented on crates/iceberg/Cargo.toml:
```diff
@@ -67,6 +67,8 @@ flate2 = { workspace = true }
 futures = { workspace = true }
 itertools = { workspace = true }
+log = { workspace = true }
+miniz_oxide = "0.8"
```
We manage dependencies version in root workspace.

emkornfield replied:
done.

liurenjie1024 commented on Cargo.toml:
```diff
@@ -40,7 +40,7 @@ rust-version = "1.87"
 [workspace.dependencies]
 anyhow = "1.0.72"
-apache-avro = { version = "0.20", features = ["zstandard"] }
+apache-avro = { version = "0.20", features = ["zstandard", "snappy"] }
```
Why we need this?

emkornfield committed:
wip

emkornfield committed:
address comments

emkornfield committed:
put parsing in table properties

emkornfield committed:
fmt

emkornfield committed:
remove unneeded tests

emkornfield committed:
fix package import

emkornfield commented on crates/iceberg/src/spec/manifest/writer.rs:
```diff
@@ -43,6 +44,7 @@ pub struct ManifestWriterBuilder {
     key_metadata: Option<Vec<u8>>,
     schema: SchemaRef,
     partition_spec: PartitionSpec,
+    compression: CompressionSettings,
```
done.

emkornfield committed:
clippy and visibility

emkornfield committed:
add missing args

emkornfield committed:
address clippy

emkornfield committed:
Fmt

emkornfield committed:
move use statements to top level

emkornfield commented on crates/iceberg/src/spec/table_properties.rs:
```diff
@@ -175,12 +195,38 @@ impl TryFrom<&HashMap<String, String>> for TableProperties {
+                let level_raw = parse_property(
+                    props,
+                    TableProperties::PROPERTY_AVRO_COMPRESSION_LEVEL,
+                    255u8,
```
maybe make a constant?

emkornfield commented:
Moving to draft until we finalize https://github.com/apache/iceberg-rust/pull/1876

emkornfield committed:
update to use compression codec

emkornfield committed:
simplify test

CTTY commented:
Is this work still active?

emkornfield commented:
Yes, sorry I've been delayed in getting this ready for a re-review.

emkornfield committed:
fixes

emkornfield committed:
wip

emkornfield committed:
error

emkornfield committed:
wip

emkornfield committed:
wip

emkornfield committed:
fix

emkornfield committed:
fixes

emkornfield committed:
fix compression test and add manifest-list round-trip assertion

- Fix Gzip(9) -> Gzip(Some(9)) in simplified compression test (rebase
  reintroduced the pre-Option form)
- Add read-back assertion to manifest-list compression test so it
  verifies the compressed avro is parseable, not just smaller

Co-authored-by: Isaac
